### ISSUE TITLE
Query about the Dataset Structure

### ISSUE THREAD

**Issue by fabrahman** [declare-lab/red-instruct#1]
Hi, thanks for the great work.

I realized that in the paper you claimed having harmful questions and their **corresponding** harmless and harmful responses. Though looking at the dataset I see the blue conversations do not correspond to the `harmful question` , they only have similar topic.

I wonder if for a given harmful question you have harmful and harmless responses? and if so can you help me obtain those? Thanks much

**Comment by Bhardwaj-Rishabh** [comment]
Thanks for your interest in our work.

There are 1,912 harmful questions. Each question is asked in a conversational format. Quoting the paper itself:

"Red-LM is an agent which seeks to gain information from an ethical bot Base-LM regarding the harmful question by subtle question answering: including generic harmless queries, asking information in pieces, and providing hypothetical and fictional scenarios rather than direct querying." (Further details are mentioned in the paper)

So not every utterance in the conversation is harmful, the agenda of the conversation is to trick Base-LM into giving harmful information to a harmful question asked in the form of conversation. First couple of questions in the conversation are generally harmless, then the model tends to ask harmful questions directly or indirectly, then based on the Base-LM's response, it may change the flow of conversation and may ask a harmless question followed by a harmful question. In the end, the main goal is to get harmful information to a main harmful question (out of those 1,912) given to it in a conversational format, each utterance of which may or may not be harmful.

Hope this clarifies.

**Comment by fabrahman** [comment]
Thanks much for getting back to me.

Do you by any chance have a subset of harmful question and harmful responses? Like jaillbreak examples?

Did you have a way of detecting which conversation round on red conversation is in fact harmful for your alignment training? or did you simply took all the red conversation for doing gradient ascent? 

Thanks

**Comment by Bhardwaj-Rishabh** [comment]
Yes, we did a gradient ascent on all the red responses while doing gradient descent on blue responses, asking the model to respond in blue-style rather than red-style (red-blue responses make a preference dataset). Consider a human ranking the responses and always choosing blue responses over red. The main goal is to minimize the impact of the CoU (conversational) prompt on the model.

Having said that, we did not filter harmful question-answers from the conversations. However, it is not at all tough, you can employ [gpt4 to annotate](https://github.com/declare-lab/red-instruct/blob/main/gpt4_as_judge.py) for you. It might be a little costly though but extremely simple.
